pyrefly/lib/state/semantic_tokens.rs

```diff
@@ -302,3 +358,86 @@ impl SemanticTokenBuilder {
         tokens
     }
 }
+
+fn collect_disabled_ranges_from_block(
+    stmts: &[Stmt],
+    sys_info: &SysInfo,
+    reachable: bool,
+    ranges: &mut Vec<TextRange>,
+) {
+    for stmt in stmts {
+        collect_disabled_ranges_from_stmt(stmt, sys_info, reachable, ranges);
+    }
+}
+
+fn collect_disabled_ranges_from_stmt(
+    stmt: &Stmt,
+    sys_info: &SysInfo,
+    reachable: bool,
+    ranges: &mut Vec<TextRange>,
+) {
+    if !reachable {
+        ranges.push(stmt.range());
+        return;
+    }
+
+    match stmt {
+        Stmt::If(if_stmt) => {
+            let mut prior_true_branch = false;
+            for (test, body) in Ast::if_branches(if_stmt) {
+                let eval = test.and_then(|expr| sys_info.evaluate_bool(expr));
+                let branch_reachable = if prior_true_branch {
+                    false
+                } else {
+                    !matches!(eval, Some(false))
+                };
+
+                collect_disabled_ranges_from_block(body, sys_info, branch_reachable, ranges);
+
+                if !prior_true_branch && matches!(eval, Some(true)) {
+                    prior_true_branch = true;
+                }
+            }
+        }
+        Stmt::FunctionDef(func) => {
+            collect_disabled_ranges_from_block(&func.body, sys_info, reachable, ranges);
+        }
+        Stmt::ClassDef(class_def) => {
+            collect_disabled_ranges_from_block(&class_def.body, sys_info, reachable, ranges);
+        }
+        Stmt::With(with_stmt) => {
+            collect_disabled_ranges_from_block(&with_stmt.body, sys_info, reachable, ranges);
+        }
+        Stmt::For(for_stmt) => {
+            collect_disabled_ranges_from_block(&for_stmt.body, sys_info, reachable, ranges);
+            collect_disabled_ranges_from_block(&for_stmt.orelse, sys_info, reachable, ranges);
+        }
+        Stmt::While(while_stmt) => {
+            let condition = sys_info.evaluate_bool(&while_stmt.test);
+            let body_reachable = reachable && condition != Some(false);
+            collect_disabled_ranges_from_block(&while_stmt.body, sys_info, body_reachable, ranges);
+            collect_disabled_ranges_from_block(&while_stmt.orelse, sys_info, reachable, ranges);
+        }
+        Stmt::Try(try_stmt) => {
+            collect_disabled_ranges_from_block(&try_stmt.body, sys_info, reachable, ranges);
+            for handler in &try_stmt.handlers {
+                let ExceptHandler::ExceptHandler(handler) = handler;
+                collect_disabled_ranges_from_block(&handler.body, sys_info, reachable, ranges);
+            }
+            collect_disabled_ranges_from_block(&try_stmt.orelse, sys_info, reachable, ranges);
+            collect_disabled_ranges_from_block(&try_stmt.finalbody, sys_info, reachable, ranges);
+        }
+        Stmt::Match(match_stmt) => {
+            for case in &match_stmt.cases {
+                collect_disabled_ranges_from_block(&case.body, sys_info, reachable, ranges);
```

Comment:
Eventually we might want an exhaustiveness check here - a wildcard case makes everything below it unreachable. Doesn't need to be in this PR